### PR TITLE
[BC-175] Isolate libp2p dependencies

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Create more explicit `p2p` rpc API's and reorganize `eth2` rpc logic in order to use these new API's and remove libp2p-specific dependencies (`Protocol`, `Connection`, `ChannelHandlerContext`, etc).  

The bulk of the changes here involve breaking up the old `RpcMessageHandler` across the new classes `RpcHandler`, `Eth2IncomingRequestHandler`, and `Eth2OutgoingRequestHandler` and consolidating rpc-method-related logic in the new `Eth2RpcMethod` class.

### CODE REVIEW

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/beaconchain/BeaconChainMethods.java [hunk 1]
```diff
@@ -25,57 +28,126 @@
...
+        false,
+        beaconBlocksByRangeHandler,
+        peerLookup);
+  }
+
+  public Collection<RpcMethod> all() {
+    return Collections.unmodifiableCollection(allMethods);
+  }
+
+  public Eth2RpcMethod<StatusMessage, StatusMessage> status() {
```
Comment (ajsutton) [inline networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/beaconchain/BeaconChainMethods.java:138]: Strictly these should probably be `get` prefixed, but it is one of those cases where it feels right to break the rule (and I'm usually a big fan of `get` prefixes...).

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcResponseCallback.java [hunk 2]
```diff
@@ -13,57 +13,36 @@
...
-    writeResponse(ctx, rpcEncoder.encodeErrorResponse(error));
-    ctx.channel().disconnect();
-  }
-
-  @SuppressWarnings("FutureReturnValueIgnored")
-  private void writeResponse(final ChannelHandlerContext ctx, final Bytes encoded) {
-    ByteBuf respBuf = ctx.alloc().buffer();
-    respBuf.writeBytes(encoded.toArrayUnsafe());
-    ctx.writeAndFlush(respBuf);
+    rpcStream.writeBytes(rpcEncoder.encodeErrorResponse(error));
```
Comment (ajsutton) [inline networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcResponseCallback.java]: Does this need `reportExceptions()`?

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PPeer.java [hunk 3]
```diff
@@ -39,8 +50,19 @@ public boolean isConnected() {
...
+  public void disconnect() {
+    connection.getNettyChannel().close();
+  }
+
+  @Override
+  public SafeFuture<RpcStream> sendRequest(
+      RpcMethod rpcMethod, final Bytes initialPayload, final RpcRequestHandler handler) {
+    RpcHandler rpcHandler = rpcHandlers.get(rpcMethod);
+    if (rpcHandler == null) {
+      throw new IllegalArgumentException("Unknown rpc method method invoked: " + rpcMethod.getId());
```
Comment (ajsutton) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/LibP2PPeer.java]:
```suggestion
      throw new IllegalArgumentException("Unknown rpc method invoked: " + rpcMethod.getId());
```

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/LibP2PRpcStream.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.artemis.networking.p2p.rpc.RpcStream;
+import tech.pegasys.artemis.util.async.SafeFuture;
+
+public class LibP2PRpcStream implements RpcStream {
```
Comment (ajsutton) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/LibP2PRpcStream.java:24]: Really like this abstraction to hide the Netty APIs.

**Review by ajsutton — APPROVED** [review]
LGTM.  Nice work.

File: networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcResponseCallback.java (hunk 2 above)
Comment (mbaxter) [reply networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/rpc/core/RpcResponseCallback.java]: yes - good call!

File: networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/RpcHandler.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+      rpcStream = new LibP2PRpcStream(ctx);
+      activeFuture.complete(this);
+    }
+
+    public RpcStream getRpcStream() {
+      return rpcStream;
+    }
+
+    @Override
+    protected void channelRead0(final ChannelHandlerContext ctx, final ByteBuf msg) {
```
Comment (cemozerr) [inline networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/RpcHandler.java:115]: is the 0 here on the method name on purpose?

Comment (mbaxter) [reply networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/libp2p/rpc/RpcHandler.java:115]: yeah - its a quirk of the API (https://netty.io/4.0/api/io/netty/channel/SimpleChannelInboundHandler.html):

> Please keep in mind that #channelRead0(ChannelHandlerContext, I) will be renamed to messageReceived(ChannelHandlerContext, I) in 5.0.

**Pull request merged**